packages/compass-global-writes/src/store/reducer.ts

```diff
@@ -1,26 +1,251 @@
+  createShardkey: {
+    isLoading: false,
+  },
 };
 
-const reducer: Reducer<RootState, Action> = (state = initialState) => {
+const reducer: Reducer<RootState, Action> = (state = initialState, action) => {
+  if (
+    isAction<SetIsManagedNamespaceAction>(
+      action,
+      GlobalWritesActionTypes.SetIsManagedNamespace
+    )
+  ) {
+    return {
+      ...state,
+      isNamespaceSharded: action.isNamespaceManaged,
+      status: !action.isNamespaceManaged
+        ? ShardingStatuses.UNSHARDED
+        : state.status,
+    };
+  }
+
+  if (
+    isAction<ShardingInProgressStartedAction>(
+      action,
+      GlobalWritesActionTypes.ShardingInProgressStarted
+    )
+  ) {
+    return {
+      ...state,
+      createShardkey: {
+        isLoading: true,
+      },
+    };
+  }
+
+  if (
+    isAction<ShardingInProgressFinishedAction>(
+      action,
+      GlobalWritesActionTypes.ShardingInProgressFinished
+    )
+  ) {
+    return {
+      ...state,
+      isNamespaceSharded: true,
+      status: ShardingStatuses.SHARDING,
+      createShardkey: {
+        isLoading: false,
+      },
+    };
+  }
+
+  if (
+    isAction<ShardingInProgressErroredAction>(
+      action,
+      GlobalWritesActionTypes.ShardingInProgressErrored
+    )
+  ) {
+    return {
+      ...state,
+      createShardkey: {
+        isLoading: false,
+      },
+    };
+  }
+
   return state;
 };
 
+export const fetchClusterShardingData =
+  (): GlobalWritesThunkAction<Promise<void>, SetIsManagedNamespaceAction> =>
+  async (
+    dispatch,
+    getState,
+    { atlasGlobalWritesService, connectionInfoRef, logger }
+  ) => {
+    if (!connectionInfoRef.current.atlasMetadata) {
+      return;
+    }
+
+    const { namespace } = getState();
+    const { clusterName, projectId } = connectionInfoRef.current.atlasMetadata;
+
+    try {
+      // Call the API to check if the namespace is managed. If the namespace is managed,
+      // we would want to fetch more data that is needed to figure out the state and
+      // accordingly show the UI to the user.
+      const isNamespaceManaged =
+        await atlasGlobalWritesService.isNamespaceManaged(namespace, {
+          projectId,
+          clusterName,
+        });
+
+      if (!isNamespaceManaged) {
+        dispatch({
+          type: GlobalWritesActionTypes.SetIsManagedNamespace,
+          isNamespaceManaged,
+        });
+        return;
+      }
+      // TODO (COMPASS-8277): Now fetch the sharding key and possible process error.
+    } catch (error) {
+      logger.log.error(
+        logger.mongoLogId(1_001_000_330),
+        'AtlasFetchError',
+        'Error fetching cluster sharding data',
+        error as Error
+      );
+      openToast('global-writes-fetch-shard-info-error', {
+        title: `Failed to fetch sharding information: ${
+          (error as Error).message
+        }`,
+        dismissible: true,
+        timeout: 5000,
+        variant: 'important',
+      });
+    }
```

Comment:
we also want to have another state like `ERROR` or something, where user does not see loading and can see the error and possibly action to refetch.